### PR TITLE
(BIDS-1611) adding withdrawal credential API endpoint

### PR DESCRIPTION
**Pull request by Eisei24**
According to the issue we want to set a max limit so that the total request time is <100ms.
Using time.Now() and time.Since() at the start and end of the function gave me pretty much the same time for limit = 1 or limit = 100. The time slowly increases when you go into the hundreds and thousands so setting the maximum to 100 seemed fine to me.

However the total time was still ~200ms, I don't know if there is any way to reduce that.

Withdrawal credentials for Prater that can be used for testing:
1 Validator: 0x00a40412f3d9bbf799b913fdd3df4dd44c69e3d44c3659159759fd046ad8eddc
10,000 Validators: 0x0100000000000000000000008f0844fd51e31ff6bf5babe21dccf7328e19fd9f

### CODE REVIEW

**Comment by recy21** [comment]
@Eisei24 this needs an index to the withdrawalcredentials column in the db, otherwise the call will never be fast.

**Review by D13ce — CHANGES_REQUESTED** [review]
As mentioned by @recy21 and @peterbitfly , an index should help to improve the speed of your query.

Additionally: The endpoint's URL and spec has changed slightly (see according BIDS).
And please see my other review comments regarding requested changes too.

File: handlers/api.go [hunk 1]
```diff
@@ -3211,6 +3201,78 @@ func insertStats(userData *types.UserWithPremium, machine string, body *map[stri
...
+		validatorindex,
+		pubkey
+	FROM validators
+	WHERE withdrawalcredentials = $1
+	LIMIT $2
+	OFFSET $3
+	`, credential, limit, offset)
+
+	if err != nil {
+		logger.Warnf("error retrieving validator data from db: %T", err)
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply handlers/api.go:3253]: Done.

File: types/api.go [hunk 2]
```diff
@@ -318,6 +318,14 @@ type Eth1TokenTxParsed struct {
 	Operator     string    `json:"operator,omitempty"`
 }
 
+type ApiWithdrawalCredentialResponse struct {
+	WithdrawalCredential string `json:"withdrawalcredential"`
+	Validators           []struct {
+		Index  uint64 `json:"index"`
+		PubKey string `json:"pubkey"`
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply types/api.go:325]: Done.

File: handlers/api.go [hunk 3]
```diff
@@ -3231,6 +3232,71 @@ func insertStats(userData *types.UserWithPremium, machine string, body *map[stri
...
+	w.Header().Set("Content-Type", "application/json")
+
+	vars := mux.Vars(r)
+	q := r.URL.Query()
+
+	credentialsOrAddressString := vars["withdrawalCredentialsOrEth1address"]
+	credentialsOrAddressString = strings.ToLower(credentialsOrAddressString)
+	credentialsOrAddress := common.FromHex(credentialsOrAddressString)
+
+	credentials := utils.AddressToWithdrawalCredentials(credentialsOrAddress)
```
Comment (D13ce) [inline handlers/api.go:3256]: There is no sanity check. Call the function with `withdrawalCredentialsOrEth1address` set to `123` and the function will just accept that, query the db with that (=meaningless traffic) and return something like
```
{"status":"OK","data":[]}
``` 
For your convenience: http://localhost:8081/api/v1/validator/withdrawalCredentials/123

Take a look at for example https://beaconcha.in/api/v1/validator/eth1/324234234

I suggest that invalid parameters are checked and if the parameters are invalid, an error with a good description is returned. Something like
_"invalid parameter provided, neither withdrawal credentials nor eth1 address"_
(don't copy&paste this please, this is just an example I did not ponder too much)

Furthermore I think its weird to have a function called `AddressToWithdrawalCredentials` that accepts pretty much anything. Based on its name, I would expect an error if I pass `123`. Instead it returns `123` and I have no information if I can work with that or not.

My suggestion would be to remove `AddressToWithdrawalCredentials` and do everything (including sanity checks, error handling and "casting") within `ApiWithdrawalCredentialsValidators` directly.
Alternatively, have something SIMILAR to `AddressToWithdrawalCredentials` that returns withdrawalCredentials and an error based on an input that is not called `address` but `withdrawalCredentialsOrAddress`. Make the sanity checks there and return `nil` as error if the return value are valid withdrawalCredentials or return `nil` as credentials and an error based on your sanity checks. I cannot come up with a good function name for that though :smile: 

Thoughts?

**Review by D13ce — CHANGES_REQUESTED** [review]

File: handlers/api.go (hunk 3 above)
Comment (Eisei24) [reply handlers/api.go:3256]: As discussed I rewrote WithdrawalCredentialsToAddress and AddressToWithdrawalCredentials to also return errors.

I also noticed that all the "0x" regexp checks were wrong so I corrected them.

File: handlers/validator.go [hunk 4]
```diff
@@ -423,14 +423,19 @@ func Validator(w http.ResponseWriter, r *http.Request) {
 			}
 
 			timeToWithdrawal := utils.GetTimeToNextWithdrawal(distance)
+			address, err := utils.WithdrawalCredentialsToAddress(validatorPageData.WithdrawCredentials)
+			if err != nil {
+				logger.WithError(err).Error("error invalid withdrawal credentials")
```
Comment (D13ce) [inline handlers/validator.go:428]: I have talked about this with @recy21 and the conclusion is that this should be a warning.

Furthermore, `WithdrawalCredentialsToAddress` does not return an address in this case. The handling below maybe should be updated to show something like "Invalid Address" or "N/A"?
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/5203a6bb9f78b97f65312d3b0e071855dba7978e/handlers/validator.go#L438

**Review by D13ce — CHANGES_REQUESTED** [review]

File: handlers/api.go [hunk 5]
```diff
@@ -3251,9 +3251,21 @@ func ApiWithdrawalCredentialsValidators(w http.ResponseWriter, r *http.Request)
 
 	credentialsOrAddressString := vars["withdrawalCredentialsOrEth1address"]
 	credentialsOrAddressString = strings.ToLower(credentialsOrAddressString)
+
+	if !utils.IsValidEth1Address(credentialsOrAddressString) &&
+		!utils.IsValidWithdrawalCredentials(credentialsOrAddressString) {
+		logger.Warn("invalid withdrawal credentials or eth1 address provided")
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply handlers/api.go:3257]: I see, done.

File: utils/utils.go [hunk 6]
```diff
@@ -518,6 +520,11 @@ func IsValidEth1Tx(s string) bool {
 	return !zeroHashRE.MatchString(s) && eth1TxRE.MatchString(s)
 }
 
+// IsValidEth1Tx verifies whether a string represents a valid eth1-tx-hash.
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply utils/utils.go:523]: Done.

File: utils/format.go [hunk 7]
```diff
@@ -479,23 +479,24 @@ func FormatHash(hash []byte, trunc_opt ...bool) template.HTML {
...
-	return credentials
+	return nil, fmt.Errorf("invalid withdrawal credentials")
 }
 
 // AddressToWithdrawalCredentials converts a valid address to withdrawalCredentials
-func AddressToWithdrawalCredentials(address []byte) []byte {
+func AddressToWithdrawalCredentials(address []byte) ([]byte, error) {
 	if IsValidEth1Address(fmt.Sprintf("%#x", address)) {
-		credentials := make([]byte, 12)
+		credentials := make([]byte, 12, 64)
```
[the remark this answers is not in the document]
Comment (Eisei24) [reply utils/format.go:494]: Ah, right. Fixed.

File: handlers/validator.go (hunk 4 above)
Comment (Eisei24) [reply handlers/validator.go]: As discussed if there is no valid address we show N/A for the withdrawal credentials in the table and don't link anywhere.

If you want to test this I used
http://localhost:8081/validator/0xadc3f47051a6fb515636121b10b201d2cfb44f9dc5c493402eefbc198fcdf472cc7f55eed21643ac278b1bf9686d4d24#withdrawals
and forced
https://github.com/gobitfly/eth2-beaconchain-explorer/blob/67a363939653acb1ed60c1dfa08ac5bf74c4e666/handlers/validator.go#L393
to true.

**Review by D13ce — APPROVED** [review]
lgtm

**Pull request merged**